# 1. SETTINGS

In [2]:
# libraries
import numpy as np
import pandas as pd
import seaborn as sns
from datetime import date
import matplotlib.pyplot as plt
from IPython.display import display
import scipy.stats

In [3]:
# pandas options
pd.set_option("display.max_columns", None)

In [4]:
# ignore warnings
import warnings
warnings.filterwarnings("ignore")

In [5]:
# garbage collection
import gc
gc.enable()

# 2. FUNCTIONS

In [6]:
### FUNCTION FOR COUNTING MISSINGS
def count_missings(data):
    total = data.isnull().sum().sort_values(ascending = False)
    percent = (data.isnull().sum() / data.isnull().count() * 100).sort_values(ascending = False)
    table = pd.concat([total, percent], axis = 1, keys = ["Total", "Percent"])
    table = table[table["Total"] > 0]
    return table

# 3. DATA IMPORT

In [7]:
# import datasets
test  = pd.read_csv("../data/raw/Challenge_20180423.csv")
cust  = pd.read_csv("../data/raw/Customer.csv")
bond  = pd.read_csv("../data/raw/Isin.csv")
markt = pd.read_csv("../data/raw/Market.csv")
macro = pd.read_csv("../data/raw/MarketData_Macro.csv")
trade = pd.read_csv("../data/raw/Trade.csv")

In [43]:
# check all datasets
display(test.head(3))
print("Test data:", test.shape)
print("------------------------------")
display(cust.head(3))
print("Customer data:", cust.shape)
print("------------------------------")
display(bond.head(3))
print("Bonds data:", bond.shape)
print("------------------------------")
display(markt.head(3))
print("Market data:", markt.shape)
print("------------------------------")
display(macro.head(3))
print("Macroeconomic data:", macro.shape)
print("------------------------------")
display(trade.head(3))
print("Trade data:", trade.shape)

PredictionIdx   DateKey  CustomerIdx  IsinIdx BuySell  CustomerInterest
0    a1e0d80784  20180423         1856    13323     Buy               NaN
1    c2cc6cc2a8  20180423         1856     9230     Buy               NaN
2    a8e94f6344  20180423         1780     9157     Buy               NaN

Test data: (484758, 6)
------------------------------


CustomerIdx                        Sector                      Subsector  \
0         2975      Banks and Intermediaries                           Bank   
1         1594  Asset Managers & Hedge Funds                            NaN   
2          399                   Corporation  Corp - Comm. & Prof. Services   

     Region   Country  
0  Americas  BARBADOS  
1  Americas   BERMUDA  
2  Americas   BERMUDA

Customer data: (3471, 5)
------------------------------


IsinIdx  TickerIdx  ActualMaturityDateKey  IssueDateKey Seniority Currency  \
0        0        238               20381231      20051129       GOV      USD   
1        1        238               20331231      20051129       GOV      USD   
2        2        238               20331231      20051129       GOV      ARS   

       ActivityGroup    Region   Activity RiskCaptain          Owner  \
0  FLOW LOCAL MARKET  AMERICAS  ARGENTINA   ARGENTINA  EMK ARGENTINA   
1  FLOW LOCAL MARKET  AMERICAS  ARGENTINA   ARGENTINA  EMK ARGENTINA   
2  FLOW LOCAL MARKET  AMERICAS  ARGENTINA   ARGENTINA  EMK ARGENTINA   

  CompositeRating IndustrySector IndustrySubgroup MarketIssue  IssuedAmount  \
0              NR     Government        Sovereign    Domestic  1.246002e+09   
1              NR     Government        Sovereign    Domestic  4.901086e+09   
2              NR     Government        Sovereign    Domestic  1.501245e+10   

  CouponType  
0   STEP CPN  
1      FIXED  
2      FIXED

Bonds data: (27411, 17)
------------------------------


IsinIdx   DateKey    Price  Yield  ZSpread
0        1  20160101  104.250  7.835    5.505
1        7  20160101  107.500  7.520    5.541
2      102  20160101  100.746  4.048    2.085

Market data: (9867747, 5)
------------------------------


DateKey       SSE      DAX  EUROSTOXX   VSTOXX  FTSE100       HSI  \
0  20150101       NaN      NaN        NaN      NaN      NaN       NaN   
1  20150102       NaN  9764.73    3139.44  26.2531  6547.80  23857.82   
2  20150105  3350.519  9473.16    3023.14  29.6236  6417.16  23721.32   

     NIKKEI  DOWJONES_INDU    SP500    VIX  FX_USD.ARS  FX_USD.AUD  \
0       NaN            NaN      NaN  19.20    0.116959     0.81620   
1       NaN       17832.99  2058.20  17.79    0.116891     0.81140   
2  17408.71       17501.65  2020.58  19.92    0.116782     0.80905   

   FX_USD.BRL  FX_USD.CAD  FX_USD.CHF  FX_USD.CNO  FX_USD.CNY  FX_USD.EUR  \
0    0.377658    0.861030    1.006264    0.161512    0.163399      1.2101   
1    0.371292    0.851136    0.998851    0.161186    0.163292      1.2008   
2    0.366757    0.850268    0.993764    0.160822    0.163265      1.1939   

   FX_USD.GBP  FX_USD.HKD  FX_USD.IDR  FX_USD.JPY  FX_USD.NOK  FX_USD.SGD  \
0     1.55775    0.128941    0.000081    0.008341    0.133451    0.756287   
1     1.53340    0.128934    0.000080    0.008308    0.131841    0.752757   
2     1.52555    0.128926    0.000079    0.008366    0.131277    0.749260   

   FX_USD.TRY  FX_USD.ZAR  MoneyMarket_ARS3M  MoneyMarket_AUD3M  \
0    0.428964    0.086457          35.230452           2.724540   
1    0.427241    0.085609          33.000000           2.736281   
2    0.429304    0.085426          32.000000           2.742523   

   MoneyMarket_CAD3M  MoneyMarket_CHF3M  MoneyMarket_CNO3M  MoneyMarket_CNY3M  \
0           1.303309          -0.063143           3.701588           3.617925   
1           1.302850          -0.063000           3.701588           3.617925   
2           1.303171          -0.063000           3.749732           3.691439   

   MoneyMarket_EUR3M  MoneyMarket_GBP3M  MoneyMarket_HKD3M  MoneyMarket_IDR3M  \
0           0.077487           0.561227           0.381829            7.17183   
1           0.077179           0.563380           0.385000            7.10000   
2           0.075252           0.563380           0.383570            7.10000   

   MoneyMarket_JPY3M  MoneyMarket_NOK3M  MoneyMarket_SGD3M  MoneyMarket_TRY3M  \
0           0.112078           1.459648           0.709986          10.001685   
1           0.112078           1.490000           0.790000          10.020000   
2           0.112140           1.450000           0.875000           9.993400   

   MoneyMarket_USD3M  MoneyMarket_ZAR3M  Swap_ARS10Y  Swap_ARS2Y  Swap_ARS5Y  \
0           0.256570           6.130798    19.989643   29.980614   21.985733   
1           0.255218           6.125000    20.000000   30.000000   22.000000   
2           0.254194           6.125000    20.000000   30.000000   22.000000   

   Swap_AUD10Y  Swap_AUD2Y  Swap_AUD30Y  Swap_AUD5Y  Swap_BRL10Y  Swap_BRL2Y  \
0     3.041150    2.398950     3.625076    2.558248          NaN         NaN   
1     3.108394    2.452905     3.692179    2.625847    11.767860   12.807338   
2     3.007227    2.423082     3.591836    2.566016    11.809585   12.736379   

   Swap_BRL5Y  Swap_CAD10Y  Swap_CAD2Y  Swap_CAD30Y  Swap_CAD5Y  Swap_CHF10Y  \
0         NaN     2.277279         NaN     2.760043    1.775859     0.409878   
1   12.202677     2.231223    1.445105     2.716551    1.748069     0.367391   
2   12.154331     2.180266    1.420899     2.668708    1.713829     0.350011   

   Swap_CHF2Y  Swap_CHF30Y  Swap_CHF5Y  Swap_CNH10Y  Swap_CNH2Y  Swap_CNH30Y  \
0         NaN     1.082563   -0.046046     3.749868    3.418686          NaN   
1         NaN     1.040147   -0.058452     3.750081    3.419989          NaN   
2   -0.263049     0.981356   -0.069792     3.700010    3.405070          NaN   

   Swap_CNH5Y  Swap_CNO10Y  Swap_CNO2Y  Swap_CNO30Y  Swap_CNO5Y  Swap_CNY10Y  \
0    3.549519     3.750081    3.419989          NaN    3.550031     3.439794   
1    3.550031     3.750081    3.419989          NaN    3.550031     3.439794   
2    3.500016     3.700010    3.405070      

Macroeconomic data: (877, 112)
------------------------------


TradeDateKey  CustomerIdx  IsinIdx BuySell  NotionalEUR  Price TradeStatus  \
0      20161207         2789     8478    Sell     653168.0    0.0     Unknown   
1      20170329         2574    14562     Buy    1656487.0    0.0     Unknown   
2      20170418         2574     4747     Buy     939673.0    0.0     Unknown   

   CustomerInterest  
0               1.0  
1               1.0  
2               1.0

Trade data: (6762021, 8)


# 4. PREPROCESSING

## 4.1. TRADE DATA

In [44]:
# check missings
count_missings(trade)

Total    Percent
Price  4617933  68.292201

In [45]:
trade.describe()

TradeDateKey   CustomerIdx       IsinIdx   NotionalEUR         Price  \
count  6.762021e+06  6.762021e+06  6.762021e+06  6.762021e+06  2.144088e+06   
mean   2.016750e+07  1.922354e+03  1.460369e+04  6.313228e+06  1.779008e+05   
std    6.768082e+03  8.579293e+02  7.963809e+03  2.712408e+08  1.318503e+06   
min    2.016010e+07  0.000000e+00  0.000000e+00 -1.485547e+08 -9.999990e+05   
25%    2.016063e+07  1.288000e+03  7.392000e+03  3.410980e+05  9.313780e+01   
50%    2.017023e+07  2.090000e+03  1.522900e+04  8.812120e+05  1.010000e+02   
75%    2.017093e+07  2.574000e+03  2.211900e+04  2.136842e+06  1.075000e+02   
max    2.018042e+07  3.470000e+03  2.739400e+04  2.000000e+11  9.999999e+06   

       CustomerInterest  
count      6.762021e+06  
mean       3.271106e-01  
std        4.691581e-01  
min        0.000000e+00  
25%        0.000000e+00  
50%        0.000000e+00  
75%        1.000000e+00  
max        1.000000e+00

In [10]:
trade["CustomerInterest"] = 1
trade["CustomerInterest"][trade["TradeStatus"] == "Holding"] = 0

In [11]:
# convert dates
trade["TradeDateKey"] = pd.to_datetime(trade["TradeDateKey"], format = '%Y%m%d')

In [12]:
### add week index

# function
def week_lag_idx(date, end_date):
    return round((end_date - date).dt.days / 7 + 0.4)

# week index
trade["WeekLag"] = week_lag_idx(trade["TradeDateKey"], date(2018, 4, 23))
test["WeekLag"] = 0

In [39]:
# aggregate weekly data
vars = ["WeekLag", "CustomerIdx", "IsinIdx", "BuySell", "CustomerInterest"]
train = trade[vars].groupby(["WeekLag", "CustomerIdx", "IsinIdx", "BuySell"], as_index = False).max()

The number of observations and bonds per customer is different in trade and test data. Trade data only contains bonds that a given customer has actually traded. In test data, for each customer, the set of bonds is only a subset of the bonds that he actually traded in the past (but not the whole set, which leads to a smaller number of observations per customer in the test data). Also, the test set is biased towards very active traders.

In [50]:
# check number of cases per customer
display(trade.groupby("CustomerIdx").IsinIdx.nunique().describe())
print("------------------------------")
display(test.groupby("CustomerIdx").IsinIdx.nunique().describe())

count     3439.000000
mean       405.832800
std       1145.875089
min          1.000000
25%          5.000000
50%         33.000000
75%        233.500000
max      14132.000000
Name: IsinIdx, dtype: float64

------------------------------


count    2495.000000
mean       97.145892
std       301.375388
min         1.000000
25%         3.000000
50%        12.000000
75%        58.000000
max      5044.000000
Name: IsinIdx, dtype: float64

In [ ]:
### ADDING MISSING COMBINATIONS

# extract all unique values
#l1 = list(trade.TradeDateKey.unique())
#l2 = list(trade.CustomerIdx.unique())
#l3 = list(trade.IsinIdx.unique())
#l4 = list(trade.BuySell.unique())

# create lists with combinations
#lp1, lp2, lp3, lp4 = pd.core.reshape.util.cartesian_product([l1, l2, l3, l4])

# convert to dataframe
#combs = pd.DataFrame(dict(TradeDateKey = lp1, CustomerIdx = lp2, IsinIdx = lp3, BuySell = lp4))

# merge missing combinations
#print(trade.shape)
#trade = trade.merge(combs, how = "right")
#print(trade.shape)

## 4.2. CUSTOMER DATA

TODO: Solve the issue of biased sampling due to the mandatory reporting before trusting the statistics!

There are 3471 unique customers in the training data of which 2495 show up in the test set. There are no unknown customers in the test set.

In [14]:
cust.CustomerIdx.nunique()

3471

In [15]:
# Compare number of test customers showing up in cust set to number of unique customers in test set
np.sum(np.in1d(test.CustomerIdx.unique(), cust.CustomerIdx.unique()))/test.CustomerIdx.nunique()

1.0

In [13]:
np.percentile(trade.groupby("CustomerIdx").size(), range(0,100,10))

array([1.0000e+00, 2.0000e+00, 5.0000e+00, 1.1000e+01, 2.5000e+01,
       5.5000e+01, 1.3600e+02, 3.6060e+02, 9.6940e+02, 3.9168e+03])

Customer information:
- 5 different sectors
- 41 subsectors
- 3 regions
- 99 countries

We can include sectors and regions directly, but may want to reduce the dimension of subsectors and countries

In [26]:
cust.columns

Index(['CustomerIdx', 'Sector', 'Subsector', 'Region', 'Country'], dtype='object')

In [28]:
cust_dummies = pd.concat([cust.CustomerIdx, pd.get_dummies(cust.Sector), pd.get_dummies(cust.Region)], axis=1)

In [30]:
cust_dummies.describe()

CustomerIdx  Asset Managers & Hedge Funds  Asset Owners  \
count  3471.000000                   3471.000000   3471.000000   
mean   1735.000000                      0.431288      0.096226   
std    1002.135719                      0.495327      0.294943   
min       0.000000                      0.000000      0.000000   
25%     867.500000                      0.000000      0.000000   
50%    1735.000000                      0.000000      0.000000   
75%    2602.500000                      1.000000      0.000000   
max    3470.000000                      1.000000      1.000000   

       Banks and Intermediaries  Corporation  Official Institution - OI  \
count               3471.000000  3471.000000                3471.000000   
mean                   0.414002     0.018438                   0.040046   
std                    0.492620     0.134550                   0.196096   
min                    0.000000     0.000000                   0.000000   
25%                    0.000000     0.000000                   0.000000   
50%                    0.000000     0.000000                   0.000000   
75%                    1.000000     0.000000                   0.000000   
max                    1.000000     1.000000                   1.000000   

          Americas  Asia Pacific  Europe, Middle East and Africa  
count  3471.000000   3471.000000                     3471.000000  
mean      0.252089      0.222126                        0.525785  
std       0.434274      0.415735                        0.499407  
min       0.000000      0.000000                        0.000000  
25%       0.000000      0.000000                        0.000000  
50%       0.000000      0.000000                        1.000000  
75%       1.000000      0.000000                        1.000000  
max       1.000000      1.000000                        1.000000

Most customers are asset managers or banks/intermediaries (possibly high volume groups?). Distribution over Americas, Europe/Africa, Asia is 53/25/22

In [40]:
train = train.merge(cust_dummies, on = "CustomerIdx", how='left')
test = test.merge(cust_dummies, on = "CustomerIdx", how='left')

In [41]:
train.columns

Index(['WeekLag', 'CustomerIdx', 'IsinIdx', 'BuySell', 'CustomerInterest',
       'Asset Managers & Hedge Funds', 'Asset Owners',
       'Banks and Intermediaries', 'Corporation', 'Official Institution - OI',
       'Americas', 'Asia Pacific', 'Europe, Middle East and Africa'],
      dtype='object')

In [42]:
train[['Asset Managers & Hedge Funds', 'Asset Owners',
       'Banks and Intermediaries', 'Corporation', 'Official Institution - OI']].mean()

Asset Managers & Hedge Funds    0.735066
Asset Owners                    0.056058
Banks and Intermediaries        0.199650
Corporation                     0.001259
Official Institution - OI       0.007966
dtype: float64

## 4.3. MACROECONOMIC DATA

In [46]:
macro.describe()

DateKey          SSE           DAX    EUROSTOXX      VSTOXX  \
count  8.770000e+02   875.000000    876.000000   876.000000  875.000000   
mean   2.016278e+07  3310.487480  11355.682289  3331.244218   20.305856   
std    9.840161e+03   413.973635   1136.826936   255.117381    5.979965   
min    2.015010e+07  2655.661000   8752.870000  2680.350000   10.678300   
25%    2.015110e+07  3085.148000  10377.267500  3069.882500   15.286350   
50%    2.016091e+07  3222.167000  11448.480000  3407.065000   19.918000   
75%    2.017071e+07  3381.246500  12299.285000  3553.512500   23.934450   
max    2.018051e+07  5166.350000  13559.600000  3828.780000   40.803000   

           FTSE100           HSI        NIKKEI  DOWJONES_INDU        SP500  \
count   876.000000    876.000000    875.000000     876.000000   876.000000   
mean   6873.943179  24726.487500  19148.118977   19709.190297  2256.015862   
std     515.046544   3406.147361   2008.348280    2709.444089   245.959213   
min    5536.970000  18319.580000  14952.020000   15660.180000  1829.080000   
25%    6401.537500  22137.215000  17348.850000   17731.567500  2076.547500   
50%    6936.395000  24113.730000  19344.150000   18330.890000  2162.430000   
75%    7327.522500  27525.542500  20261.175000   21487.745000  2438.467500   
max    7778.640000  33154.120000  24124.150000   26616.710000  2872.870000   

              VIX  FX_USD.ARS  FX_USD.AUD  FX_USD.BRL  FX_USD.CAD  FX_USD.CHF  \
count  877.000000  877.000000  877.000000  877.000000  877.000000  877.000000   
mean    14.840485    0.075868    0.757052    0.302329    0.771900    1.025955   
std      4.483849    0.022469    0.026942    0.027013    0.028120    0.028106   
min      9.140000    0.043478    0.685800    0.241255    0.686460    0.970968   
25%     11.625000    0.058997    0.739300    0.287051    0.752672    1.006441   
50%     13.770000    0.066094    0.759850    0.307342    0.770090    1.025247   
75%     16.640000    0.104482    0.775550    0.317531    0.792801    1.040799   
max     40.740000    0.116959    0.823700    0.389226    0.861030    1.169967   

       FX_USD.CNO  FX_USD.CNY  FX_USD.EUR  FX_USD.GBP  FX_USD.HKD  FX_USD.IDR  \
count  877.000000  877.000000  877.000000  877.000000  877.000000  877.000000   
mean     0.153128    0.153606    1.127517    1.391046    0.128601    0.000075   
std      0.005431    0.006053    0.050020    0.109302    0.000490    0.000002   
min      0.143740    0.143657    1.038800    1.204400    0.127389    0.000068   
25%      0.149432    0.149410    1.090700    1.296750    0.128162    0.000074   
50%      0.152718    0.152917    1.119600    1.391850    0.128887    0.000075   
75%      0.157803    0.157803    1.160600    1.496100    0.128966    0.000076   
max      0.161512    0.163720    1.250900    1.588050    0.129034    0.000081   

       FX_USD.JPY  FX_USD.NOK  FX_USD.SGD  FX_USD.TRY  FX_USD.ZAR  \
count  877.000000  877.000000  877.000000  877.000000  877.000000   
mean     0.008848    0.122075    0.729053    0.317650    0.075063   
std      0.000494    0.004979    0.018090    0.046718    0.006563   
min      0.007960    0.111581    0.689489    0.232794    0.059245   
25%      0.008394    0.118018    0.714082    0.275756    0.070482   
50%      0.008870    0.121448    0.732681    0.329707    0.074733   
75%      0.009154    0.126018    0.741977    0.346512    0.080488   
max      0.010008    0.137120    0.766019    0.438597    0.088519   

       MoneyMarket_ARS3M  MoneyMarket_AUD3M  MoneyMarket_CAD3M  \
count         877.000000         877.000000         877.000000   
mean           33.804797           1.982505           1.061660   
std            43.144767           0.255217           0.287314   
min            14.850000           1.688565           0.714657   
25%            20.800000           1.754142           0.889615   
50%            23.000000           1.952123           0.932760   
75%            26.000000           2.170082           1.282437   
max           259.00000

## 4.4. BONDS DATA

Check that bonds after maturity are not tradeable. If this is correct, the easiest solution would be to manually correct these to 0 in model predictions

In [52]:
bond.describe(include="all")

IsinIdx     TickerIdx  ActualMaturityDateKey  IssueDateKey  \
count   27411.000000  27411.000000           2.741100e+04  2.741100e+04   
unique           NaN           NaN                    NaN           NaN   
top              NaN           NaN                    NaN           NaN   
freq             NaN           NaN                    NaN           NaN   
mean    13705.000000   1748.038598           2.023912e+07  2.013163e+07   
std      7913.018451   1028.719783           8.515510e+04  4.165186e+04   
min         0.000000      0.000000           2.015043e+07  1.955010e+07   
25%      6852.500000    855.000000           2.019042e+07  2.012013e+07   
50%     13705.000000   1753.000000           2.022013e+07  2.014062e+07   
75%     20557.500000   2647.000000           2.025091e+07  2.016071e+07   
max     27410.000000   3548.000000           2.167020e+07  2.018053e+07   

       Seniority Currency ActivityGroup    Region Activity     RiskCaptain  \
count      27411    27411         27411     27411    27411           27411   
unique         9       23             3         8       16              38   
top          SEN      USD      FLOW G10  AMERICAS  HG CASH  HG CASH NONFIN   
freq       21233    16358         16552     10732     7928            5608   
mean         NaN      NaN           NaN       NaN      NaN             NaN   
std          NaN      NaN           NaN       NaN      NaN             NaN   
min          NaN      NaN           NaN       NaN      NaN             NaN   
25%          NaN      NaN           NaN       NaN      NaN             NaN   
50%          NaN      NaN           NaN       NaN      NaN             NaN   
75%          NaN      NaN           NaN       NaN      NaN             NaN   
max          NaN      NaN           NaN       NaN      NaN             NaN   

                           Owner CompositeRating IndustrySector  \
count                      27411           27411          27406   
unique                       105              29             13   
top     US NONFIN 5Y AND IN CASH              NR      Financial   
freq                        1714            4874          10046   
mean                         NaN             NaN            NaN   
std                          NaN             NaN            NaN   
min                          NaN             NaN            NaN   
25%                          NaN             NaN            NaN   
50%                          NaN             NaN            NaN   
75%                          NaN             NaN            NaN   
max                          NaN             NaN            NaN   

           IndustrySubgroup MarketIssue  IssuedAmount CouponType  
count                 27406       27394  2.741100e+04      27411  
unique                  338          14           NaN          6  
top     Commer Banks Non-US    Euro mtn           NaN      FIXED  
freq                   3715        7374           NaN      24105  
mean                    NaN         NaN  2.044049e+09        NaN  
std                     NaN         NaN  5.274043e+10        NaN  
min                     NaN         NaN  0.000000e+00        NaN  
25%                     NaN         NaN  5.000000e+08        NaN  
50%                     NaN         NaN  7.500000e+08        NaN  
75%                     NaN         NaN  1.200000e+09        NaN  
max                     NaN         NaN  5.400000e+12        NaN

In [54]:
bond_dummies = pd.concat([bond.IsinIdx, 
                          pd.get_dummies(bond.ActivityGroup), 
                          pd.get_dummies(bond.CompositeRating)], axis=1)

In [55]:
train = train.merge(bond_dummies, on = "IsinIdx", how='left')
test = test.merge(bond_dummies, on = "IsinIdx", how='left')

## 4.5. MARKET DATA

# 5. DATA EXPORT

In [17]:
# export CSV
train.to_csv("../data/prepared/train.csv", index = False, float_format = "%.8f")
test.to_csv("../data/prepared/test.csv",   index = False, float_format = "%.8f")